File: strategy_pattern/strategy_pattern.py
# -*- coding: utf-8 -*-


class Building(object):

    def __init__(self, action, work):
        self.do_action = action()
        self.do_work = work()

    def action(self):
        self.do_action.action()

    def work(self):
        self.do_work.work()

    def livehuman(self):
        print "people live here!!"


class Officetel(Building):

    def __init__(self):
        Building.__init__(self, Sleeping, Nowork)
        print "it's just normal house"


class Policeoffice(Building):

    def __init__(self):
        Building.__init__(self, Catchcriminal, Hardwork)
        print "it's policeoffice!!"


class Firestation(Building):

    def __init__(self):
        Building.__init__(self, Stopfire, Easywork)
        print "it's fire station!!"


class Action():

    def action(self):
        pass


class Stopfire(Action):

    def action(self):
        print "we get fire off !!"


class Catchcriminal(Action):

    def action(self):
        print "NYPD!!"


class Sleeping(Action):

    def action(self):
        print "Zzzz....."


class Work():

    def work(self):
        pass


class Nowork(Work):

    def work(self):
        print "Noworking"


class Hardwork(Work):

    def work(self):
        print "so hard!!"


class Easywork(Work):

    def work(self):
        print "easy work :)"

if __name__ == "__main__":

    officetel = Officetel()
    officetel.action()
    officetel.work()
    officetel.livehuman()

    print "\n**********************"
    policeoffice = Policeoffice()
    policeoffice.action()
    policeoffice.work()
    policeoffice.livehuman()
    policeoffice.do_action = Sleeping()
    policeoffice.action()

    print "\n**********************"

    firestation = Firestation()
    firestation.action()
    firestation.work()
    firestation.livehuman()
    firestation.do_action = Sleeping()
    firestation.action()

File: strategy_pattern/strategy_pattern.go
package main

import (
	"fmt"
)

type Building struct {
	do_action Actioner
	do_work   Worker
}

func (b *Building) SetAction(a Actioner) {
	b.do_action = a
}

func (b *Building) SetWork(w Worker) {
	b.do_work = w
}

func (b *Building) live() {
	fmt.Println("im alive here!!")
}

type Actioner interface {
	action()
}

type Stopfire struct{}

func (this *Stopfire) action() {
	fmt.Println("we get fire off !!")
}

type Catchcriminal struct{}

func (this *Catchcriminal) action() {
	fmt.Println("NYPD!!")
}

type Sleeping struct{}

func (this *Sleeping) action() {
	fmt.Println("Zzzz.......")
}

type Worker interface {
	work()
}

type Nowork struct{}

func (this *Nowork) work() {
	fmt.Println("Noworking")
}

type Hardwork struct{}

func (this *Hardwork) work() {
	fmt.Println("so hard!!")
}

type Easywork struct{}

func (this *Easywork) work() {
	fmt.Println("easy work :) ")
}

func Officetel() *Building {
	return &Building{new(Sleeping), new(Nowork)}
}

func Policeoffice() *Building {
	return &Building{new(Catchcriminal), new(Hardwork)}
}

func Firestation() *Building {
	return &Building{new(Stopfire), new(Easywork)}
}

func main() {

	officetel := Officetel()
	officetel.do_action.action()
	officetel.do_work.work()
	officetel.SetAction(new(Sleeping))
	officetel.do_action.action()
	officetel.live()

	fmt.Println("***********************")
	policeoffice := Policeoffice()
	policeoffice.do_action.action()
	policeoffice.do_work.work()
	policeoffice.SetAction(new(Sleeping))
	policeoffice.do_action.action()
	policeoffice.live()

	fmt.Println("***********************")
	firestation := Firestation()
	firestation.do_action.action()
	firestation.do_work.work()
	firestation.SetWork(new(Nowork))
	firestation.do_work.work()
	firestation.live()
}
